### ISSUE TITLE
[Feature] Implement hotspot gateway

### ISSUE THREAD

**Issue by Buster14** [tomaae/homeassistant-mikrotik_router#200]
Hi, i'm trying to differentiate HA trigger between A, P, S and D / H status in Hotspot > Hosts status.
The reason for this is that there's some clients that only connects and not logged in to my hotspot server, and that messes up the automation. My AP is unifi and my radius server and hotspot server is mikrotik. Currently using unifi counter as trigger but of course i'm getting alot of false positive trigger due to above issue
Can this be done with this integration?
How would i create a sensor then?

**Comment by tomaae** [comment]
Sorry, I dont know what you mean. I'm not too familiar with mikrotik wifi and not at all with hotspot.
Can you go more into detail?

**Comment by Buster14** [comment]
So in mikrotik hotspot, when a client is connected to the Access point, mikrotik flags them with one of these 5 flags :

S - static
H - DHCP
D - dynamic,
A - authorized
P - bypassed 

When client first connected to the wifi AP, it's flagged as H, since mikrotik will give it iP with DHCP
After logged in to the radius server,
(or for simplicity sake, using a bought wifi voucher and then enter username and password.)
It will be flagged as H & A since it's Authorized already.

The problem is in my case, there are some people who are connected but not logged in. a.k.a only H
While for HA trigger i need H&A counter sensor,
the current unifi counter only counts all wifi clients, H included, so i get alot of false positives.

So basically what i need is to query mikrotik IP>hotspot>host for H&A clients. (not H OR A)
and then present them as a counter sensor.

For example : 

P      userA
H      userB
H&A user C
P       user D
H&A  user E

your counter sensor would output 2 since there are 2 H&A users.

I've been trying to find the API syntax for this, but i think you should be more experienced than me?
I found some related examples but i think they are CLI

[H&A clients. (not H OR A)](https://forum.mikrotik.com/viewtopic.php?t=24509)

**Comment by tomaae** [comment]
Ok, so if I understand it right, you need to have these flags exposed to HA. That should be doable.
Could you run integration in debug mode for a while to let few people connect via hostspot? I'm only interested in wireless lines.
It will be better to see it on live system, so information is accurate.

**Comment by Buster14** [comment]
Yes, basically just a counter for "A" flag 

Here you go, i logged in a device with voucher between 1638-1639, so i shorten the logs

https://drive.google.com/file/d/1YKnaTRmewCiVFrBmIVZOo_AKI-_I1jwX/view?usp=sharing

**Comment by tomaae** [comment]
hmm, I may be misunderstanding what you need. 
You are not actually using hotspot, but just hotspot gateway on mikrotik, right?
if so, this is something that is not implemented at all in integration.
Could you point me to a place that explains purpose of hotspot gateway and how it works? I need to first understand it to create proper implementation.

I know its bit too much, but could you give me outputs of following? You can of course mask sensitive info.
```
/ip hotspot print terse
/ip hotspot active print terse
/ip hotspot host print terse
/ip hotspot user print terse
/ip hotspot profile print terse
```

**Comment by Buster14** [comment]
Hmm sorry if i cant explain it better, i'm not a mikrotik expert as well, 
But from my understanding, it's some kind of gateway / server,
it stores the configs of the hotspot server desired, e.g 
this page explains it well
https://nds.id/en/interesting-features-and-mikrotik-hotspot-settings/

Of course, no problem at all, let me know if you need anything else :
```
[buster@rc.01.0150.CEMARA-NET] > ip hotspot print terse
 0   name=hotspot1 interface=ether8 address-pool=hs-pool-14 profile=hsprof1 idle-t
imeout=5m keepalive-timeout=none login-timeout=none addresses-per-mac=2 ip-of-dns-
name=192.168.88.1 proxy-status=running 

[buster@rc.01.0150.CEMARA-NET] > ip hotspot active print terse

[buster@rc.01.0150.CEMARA-NET] > ip hotspot host print terse
 0  P comment=AP AC-Lite mac-address=F4:f2:BF:0B:D4:3F address=192.168.88.253 to-a
ddress=192.168.88.253 server=hotspot1 uptime=1d12h13m39s idle-timeout=5m found-by=
UDP :52959 -> 168.119.238.107:123 
 1  P mac-address=48:E7:DA:68:3C:hB address=192.168.88.224 to-address=192.168.88.2
24 server=hotspot1 uptime=1d12h13m38s idle-timeout=5m found-by=ARP reply to 192.16
8.88.1 
 2  P mac-address=48:55:e9:02:FF:3B address=192.168.88.245 to-address=192.168.88.2
45 server=hotspot1 uptime=1d12h13m25s idle-timeout=5m found-by=ARP reply to 192.16
8.88.1 

[buster@rc.01.0150.CEMARA-NET] > ip hotspot user print terse
 0 * comment=counters and limits for trial users name=default-trial uptime=0s byte
s-in=0 bytes-out=0 packets-in=0 packets-out=0 
 1   name=admin password=admin profile=default uptime=0s bytes-in=0 bytes-out=0 pa
ckets-in=0 packets-out=0 

[buster@rc.01.0150.CEMARA-NET] > ip hotspot profile print terse
 0 * name=default hotspot-address=0.0.0.0 dns-name="" html-directory=hotspot html-
directory-override="" rate-limit="" http-proxy=0.0.0.0:0 smtp-server=0.0.0.0 login
-by=cookie,http-chap http-cookie-lifetime=3d split-user-domain=no use-radius=no 
 1   name=hsprof1 hotspot-address=192.168.88.1 dns-name=hs.cdddd.net html-directo
ry=hotspot html-directory-override="" rate-limit="" http-proxy=0.0.0.0:0 smtp-serv
er=0.0.0.0 login-by=cookie,http-chap http-cookie-lifetime=3d split-user-domain=no 
use-radius=yes radius-accounting=yes radius-interim-update=received nas-port-type=
wireless-802.11 radius-default-domain="" radius-location-id="" radius-location-nam
e="" radius-mac-format=XX:XX:XX:XX:XX:XX 
```

**Comment by tomaae** [comment]
thanks I will go through it.

oh, so its that thing that hotels usually have? where you connect to open public wifi and can only access login page where you must authenticate to gain internet access?

**Comment by Buster14** [comment]
Yes, 
although they dont always use mikrotik for their hotspot server. Could be pfsense, etc.

**Comment by tomaae** [comment]
thats fine, I just needed to understand what it actually does. thanks.
I will go through the article and data you sent me and see how to implement this. my guess is that this could be useful to more people.
if not, at least I learn something new :)

**Comment by Buster14** [comment]
Hi, would you like a freelancer project for this? I think it's better that way. I can compensate your effort as well.

**Comment by tomaae** [comment]
There is really no need to. I do not require payment for implementing new features. Specially when I dont need new hardware to work on it.
I usually implement new features in larger bundles, since that is more time effective and I preffer to focus on one thing at time.
I split my time between several open source projects and other projects not related to github. It may take some time, but I implement mostly everything that makes sense.
This may be rarely used feature, but I see no reason to skip on it. It will be certainly useful for some people.
So dont worry, I will work on it. Just not right away.
Is this something you actually need urgently?

**Comment by Buster14** [comment]
Well, it's not that urgent,  
but I'm just worried it might be abandoned if its too difficult or take months to complete..  and with freelancer at least there's a clear time goal and bugs get solved quicker.
If you look at the codes do you think it's difficult to implement?

**Comment by tomaae** [comment]
Not sure yet. Since states via API are not documented, it may be tough without setting up captive portal myself.
CLI gives these data differently then API, since its not intended to be visually recognizable.
Its just one of those features I never used.
Once I get example data from API, it will be simple enough.

**Comment by Buster14** [comment]
The API that explains querying should be covered here :

https://help.mikrotik.com/docs/display/ROS/API#API-Queries
Some examples :
https://forum.mikrotik.com/viewtopic.php?f=2&t=72298

I think the query code might be like this, but it needs to be tested.

```
/ip/hotspot/hosts/print
?authenticated=yes
```

Also there is no count so it has to be counted from the python side i guess...

**Comment by tomaae** [comment]
Thats not what I mean.

Lets do it like this to speed it up since you seem to need it fast. 
Can you replace mikrotik controller temporary with this one? Remember to backup original, so you can put it back.
It will print all hotspot data as seen though API. It is setup as warning output in HA log and nothing will be masked. You can mask things you consider sensitive of course.
Just make sure you capture states A, P, S and D / H so we know how exactly are they reported using API.
If you can get this data, it will be easy for me to add this.
[mikrotik_controller.zip](https://github.com/tomaae/homeassistant-mikrotik_router/files/8762184/mikrotik_controller.zip)

**Comment by Buster14** [comment]
https://drive.google.com/file/d/1V6kooD1oIumN2SNXVH_uqJJBXZVFlye-/view?usp=sharing
Just wanna make sure i dont miss anything, 
i replaced mikrotik.py with yours.
and i logged in using voucher with a phone so that it's authenticated.
That's good?

I logged in with mac address ending in 12:34:56, so you should search for that

**Comment by tomaae** [comment]
Started as:
'DHCP': True, 'authorized': False, 'bypassed': False
Then changed to:
'DHCP': True, 'authorized': True, 'bypassed': False

That means we are still missing S/D flags. Are they relevant? If so, we still need to look for those.

**Comment by Buster14** [comment]
Well for my case i only need count the authorized clients, so its not relevant.. For static and dynamic, shouldn't it be the same false/true flag?
let me know if there is anything else needed

**Comment by tomaae** [comment]
If that is the case, I wont need any more data from mikrotik.

Those bool flags are not present at all. Maybe they will appear when relevant, but since this is not documented, I have no idea.
From what I understand, only authorized and bypassed values are relevant values to captive portal.

I want to add this as optional feature and to add these values to tracked hosts. That should make it universal.
You said that you want to use a counter. Are you sure thats the best? Since if someone connects and other person disconnects between queries, your automation wont get triggered.

**Comment by Buster14** [comment]
Well never knew it would be that complex.. 
i'm currently using unifi counter as trigger

https://github.com/clyra/unifics

For identifying connected wifi clients (the one i told you earlier thats not 100% ideal cause it also counts unauthenticated clients)

And so far its been working fine as a trigger i think

**Comment by Buster14** [comment]
So during automation,
 A connected and authenticated
B was authenticated, then disconnected, automation is not triggered, is that what you mean? 
i havent done much automation yet so still trying to wrap my head around some of its concepts

**Comment by tomaae** [comment]
ok, try to explain what are you using that automation for. we could find a good solution.
I assume you dont care about specific clients, since you are using counter.
In case you need to know about any clients connecting (and authenticating), best could be to create a text sensor which would track last time client authenticated. Automation can just be tied to any value change for that sensor.

Both specific and any would work with implementation I'm thinking about. It would require a template sensor, but thats easy to create.

**Comment by Buster14** [comment]
It's not that complex, there are 2 automations 
First automation, whenever the number of authorized clients reaches X, turn on devices like lighting, aircon, fan, etc.
And the second automation is to turn off those devices whenever the number of authorized clients is less than X,
i am also combining both with frigate motion detection to ensure more accuracy.
Wouldn't a simple counter work for this automation?

**Comment by tomaae** [comment]
ah yes, for something like that, you only need number of people in area.
Do you also use client tracking or traffic for this integration?

**Comment by Buster14** [comment]
No, traffic wont be necessary once client is authorized, thats why its important to filter out authorized and non-authorized clients.

**Comment by tomaae** [comment]
oh, that was unrelated to this feature. just thinking about UI implementation already.
Ideally, I would want it assigned to hosts, since it makes most sense.
But separate list is also an option, tho not sure I want to duplicate hosts like that. That may be really bad design.

**Comment by tomaae** [comment]
just for info, this is not yet ready for your use case, since you dont use tracker or client traffic.
I need to think about how to expose this properly outside of individual hosts.

**Comment by Buster14** [comment]
I  see, is this tracker and client traffic available in your mikrotik integration? I only see basic config when i press the integration.
I cannot find the option to set tracker and traffic..
And you only need logs when both tracker and traffic is active for a hotspot client right?

**Comment by tomaae** [comment]
yes, it is all in 2nd config page. most data this integration provides is disabled by default.
I dont need any logs, that was just for your information. You dont need to do anything.

**Comment by Buster14** [comment]
Anything i can help to solve this?

**Comment by tomaae** [comment]
Nope, I just didnt had time to move it forward yet.
Dont worry, you will have this feature.

**Comment by tomaae** [comment]
I just got to it, can you test if it works for you? I dont have way of testing this.

**Comment by Buster14** [comment]
Thanks, i will try to test it tommorow / monday cause its sunday tmw.
So i just need to update it, and there should be a counter entity for authenticated user right?

**Comment by tomaae** [comment]
you need to install master. I have not released recent changes yet.
yes, its counter for captive portal authenticated users

**Comment by Buster14** [comment]
Hmm doesn't seem to pick up the authenticated client. I tested it twice to make sure. The sensor count stays at zero.
Here's the log : 

https://drive.google.com/file/d/1Qxd7vU_O2TD3RODdz6xm8IkLI9aEIVGa/view?usp=sharing

**Comment by tomaae** [comment]
> 

you didnt enabled captive portal in settings. its not queried.

**Comment by Buster14** [comment]
You're referring to the 2nd config page right?
How do i enable it? I only see 1 page in integration > mikrotik router > configure

**Comment by tomaae** [comment]
> You're referring to the 2nd config page right? How do i enable it? I only see 1 page in integration > mikrotik router > configure

just click submit and you will get to 2nd page

**Comment by Buster14** [comment]
![sensor](https://user-images.githubusercontent.com/13533437/175920150-0f97a21e-651a-412c-8089-8a1203b85a80.JPG)

Is it supposed to look like this?
Btw i'm only replcaing files inside "custom component/ mikrotik router"

**Comment by tomaae** [comment]
you have probably not replaced all files, or they are held in browser cache. nothing to be worried about.

**Comment by Buster14** [comment]
So it ticked all of them, but "sensor.mikrotik_captive_portal_clients" stays at zero even with 1 client logged in and authenticated.
Log:
https://drive.google.com/file/d/1S4QfCNJzHdeLP63BJ0UHoAcCWCEfQvss/view?usp=sharing

**Comment by tomaae** [comment]
I can only see authorized: false coming from mikrotik. are you sure?
There are always just 4 bypassed ones.

**Comment by Buster14** [comment]
Yes, i'm quite sure there were 1 authenticated client.
yes its correct there are 4 bypassed ones for HA devices.
btw about the 2nd config page, is it correct like that cause i dont see anything mentioning captive?

**Comment by tomaae** [comment]
it is not captured in debug then. try to capture it so I can see what is going on.
missing text it missing translation string. try to clear your browser cache. if it does not work, uninstall and reinstall master version of the integration.
You may have to go into the custom folder directly and remove anything left over after uninstalling it, since you were changing things there manually.

**Comment by Buster14** [comment]
Well thats strange, it still wont pickup the client
Can you elaborate how to install a non-released code like this?
So this is what i did : 
- delete mikrotik router integration 
- in HACS uninstall mikrotik router integration
- makes sure there are no remaining files in that folder
- install master version, i noticed it's not latest, there is no captive word at all in mikrotikcontroller.py
- so i downloaded all with githubzip and replace all contents of custom components/ mikrotik router one by one using file editor.
- replaced "en.json" in translations
- Clear cache
- Test a client login

Did i miss anything?
Logs :
https://drive.google.com/file/d/1PIYwfD4LLoaZNAB6lEodenD8F-Q6-UJE/view?usp=sharing

**Comment by tomaae** [comment]
you are actually getting an error: 
```
2022-06-29 17:37:02 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/mikrotik_router/mikrotik_controller.py", line 586, in force_update
    await self.async_update()
  File "/config/custom_components/mikrotik_router/mikrotik_controller.py", line 635, in async_update
    await self.async_process_host()
  File "/config/custom_components/mikrotik_router/mikrotik_controller.py", line 2041, in async_process_host
    self.data["host"][uid]["authorized"] = self.data["hostspot_host"][
KeyError: 'authorized'
```
fixed that, try now.

**Comment by tomaae** [comment]
> Still wont capture it,
> 
> https://drive.google.com/file/d/1PIYwfD4LLoaZNAB6lEodenD8F-Q6-UJE/view?usp=sharing

found one more typo, its now fixed in master.
but there are still no authorized hosts connected to your captive portal. this time just 3 bypassed hosts with long uptime.

**Comment by Buster14** [comment]
Sorry for the late reply, it finally works! :)
Let me test it for few days with multiple clients to make sure there are no issues.. i'll let you know soon.
Thanks for this

**Comment by tomaae** [comment]
sure thing

**Comment by Buster14** [comment]
Hi, can you push the update? i think it's working properly. 
It's not integrated yet in the latest code isnt it? I restored from a backup and installed this but there's still no "captive" in mikrotik controller py

**Comment by tomaae** [comment]
it is in master only. there are still few things to be done before next update.
